File: strings/text-justification/text-justification.go
package textjustification

func fullJustify(words []string, maxWidth int) []string {
	res := []string{}

	n := len(words)

	left := 0

	for left < n {
		l := len(words[left])

		right := left + 1

		for right < n && (l+len(words[right])+right-left) <= maxWidth {
			l += len(words[right])
			right++
		}

		extra := maxWidth - l

		if right-left == 1 || right == n {
			// left justify
			res = append(res, leftJustify(words, left, right, extra))
		} else {
			// mid justify
			res = append(res, midJustify(words, left, right, extra))
		}

		left = right
	}

	return res
}

func leftJustify(words []string, left int, right int, extra int) string {
	rightSpaces := extra - (right - left - 1)

	chars := []byte{}

	chars = append(chars, words[left]...)

	for i := left + 1; i < right; i++ {
		chars = append(chars, (" " + words[i])...)
	}

	for i := 0; i < rightSpaces; i++ {
		chars = append(chars, ' ')
	}
	return string(chars)
}

func midJustify(words []string, left int, right int, extra int) string {
	boundaries := right - left - 1
	spaces := extra / boundaries
	extraSpaces := extra % boundaries

	chars := []byte{}

	chars = append(chars, words[left]...)

	for i := left + 1; i < right; i++ {
		spacesToApply := spaces

		if extraSpaces > 0 {
			spacesToApply++
			extraSpaces--
		}

		for j := 0; j < spacesToApply; j++ {
			chars = append(chars, ' ')
		}

		chars = append(chars, words[i]...)
	}
	return string(chars)
}

File: strings/text-justification/text-justification.java
class Solution {
  public List<String> fullJustify(String[] words, int maxWidth) {
    List<String> res = new ArrayList<>();

    int left = 0;
    int n = words.length;

    while (left < n) {
      int len = words[left].length();

      int right = left + 1;

      while (right < n && (len + words[right].length() + (right - left)) <= maxWidth) {
        len += words[right].length();
        right++;
      }

      int extra = maxWidth - len;

      if (right - left == 1 || right == n) {
        // left justify
        res.add(leftJustify(words, left, right, extra));
      } else {
        // mid justify
        res.add(midJustify(words, left, right, extra));
      }

      left = right;
    }

    return res;
  }

  private String leftJustify(String[] words, int left, int right, int extra) {
    int rightSpaces = extra - (right - left - 1);
    StringBuilder sb = new StringBuilder(words[left]);

    for (int i = left + 1; i < right; i++) {
      sb.append(" " + words[i]);
    }

    for (int i = 0; i < rightSpaces; i++) {
      sb.append(" ");
    }

    return sb.toString();
  }

  private String midJustify(String[] words, int left, int right, int extra) {
    int boundaries = right - left - 1;
    int spaces = extra / boundaries;
    int extraSpaces = extra % boundaries;

    StringBuilder sb = new StringBuilder(words[left]);

    for (int i = left + 1; i < right; i++) {
      int spacesToApply = spaces;

      if (extraSpaces > 0) {
        spacesToApply++;
        extraSpaces--;
      }

      for (int j = 0; j < spacesToApply; j++) {
        sb.append(" ");
      }

      sb.append(words[i]);
    }
    return sb.toString();
  }
}
